packages/gatsby/src/utils/map-pages-to-static-query-hashes.ts

```diff
@@ -0,0 +1,174 @@
+        (result, terminalNode) => result || m?.resource?.includes(terminalNode),
+        false
+      )
+
+      // Exit if we don't have any reasons or we have reached a possible terminal node
+      if (!hasReasons || isTerminalNode) {
+        return result
+      }
+
+      // These are non terminal dependents and hence modules that need
+      // further upward traversal
+      const nonTerminalDependents = m.reasons
+        .filter(r => {
+          const dependentModule = r.module
+          const isTerminal = terminalNodes.reduce(
+            (result, terminalNode) =>
+              result || dependentModule?.resource?.includes(terminalNode),
+            false
+          )
+          return !isTerminal
+        })
+        .map(r => r.module)
+        .filter(Boolean)
+
+      const uniqDependents = uniqBy(nonTerminalDependents, (d: any) =>
+        d?.identifier()
+      )
+
+      for (const x of uniqDependents) {
+        result.add(x.resource)
+
+        if (x?.resource?.includes(`gatsby-browser.js`)) {
+          globalStaticQueries.add(staticQueryModuleComponentPath)
+        }
+        getDepsFn(x)
+      }
+
+      return result
+    }
+
+    return getDepsFn(mod)
+  }
+
+  const mapOfStaticQueryComponentsToDependants = new Map()
+
+  // For every known static query, we get its dependents.
+  staticQueryComponents.forEach(({ componentPath }) => {
+    const staticQueryComponentModule = modules.find(
+      m => m.resource === componentPath
+    )
+
+    const dependants = staticQueryComponentModule
+      ? getDeps(staticQueryComponentModule)
+      : new Set()
+
+    mapOfStaticQueryComponentsToDependants.set(componentPath, dependants)
+  })
+
+  const mapOfComponentsToStaticQueryHashes = mapComponentsToStaticQueryHashes(
+    staticQueryComponents
+  )
+
+  const globalStaticQueryHashes: string[] = []
+
+  globalStaticQueries.forEach(q => {
+    const hash = mapOfComponentsToStaticQueryHashes.get(q)
+    if (hash) globalStaticQueryHashes.push(hash)
+  })
+
+  // For every known page, we get queries
+  const mapOfTemplatesToStaticQueryHashes = Array.from(components).reduce(
+    (map, [page]) => {
+      const staticQueryHashes = [...globalStaticQueryHashes]
+
+      // Does this page contain an inline static query?
+      if (mapOfComponentsToStaticQueryHashes.has(page)) {
+        const hash = mapOfComponentsToStaticQueryHashes.get(page)
+        if (hash) staticQueryHashes.push(hash)
+      }
+
+      // Check dependencies
+      mapOfStaticQueryComponentsToDependants.forEach(
+        (setOfDependants, staticQueryComponentPath) => {
+          if (setOfDependants.has(page)) {
+            const hash = mapOfComponentsToStaticQueryHashes.get(
+              staticQueryComponentPath
+            )
+            if (hash) staticQueryHashes.push(hash)
+          }
+        }
+      )
+      map.set(page, staticQueryHashes)
+      return map
+    },
+    new Map()
+  )
+
+  return mapOfTemplatesToStaticQueryHashes
+}
```

Comment:
This file is gnarly! You did a great job documenting it, and it's relatively easy to read.

My main worry is there is a lot of looping going on here. Have you measured this function at a site of decent scale to see how heavy it is?

Reply:
All loops are function of how many static queries there are and how many independent templates (and not pages) there are in the site so we should fine.